src/computability/DFA.lean

```diff
@@ -43,4 +44,96 @@ def eval := M.eval_from M.start
 def accepts : language α :=
 λ x, M.eval x ∈ M.accept
 
+lemma mem_accepts (x : list α) : x ∈ M.accepts ↔ M.eval_from M.start x ∈ M.accept := by refl
+
+lemma eval_from_of_append (start : σ) (x y : list α) :
+  M.eval_from start (x ++ y) = M.eval_from (M.eval_from start x) y :=
+list.foldl_append _ _ x y
+
+lemma eval_from_split [fintype σ] (x : list α) (s t : σ) :
+  fintype.card σ + 1 ≤ x.length →
+    M.eval_from s x = t →
+      ∃ q a b c,
+        x = a ++ b ++ c ∧
+        a.length + b.length ≤ fintype.card σ + 1 ∧
+        b ≠ [] ∧
+        M.eval_from s a = q ∧
+        M.eval_from q b = q ∧
+        M.eval_from q c = t :=
```

Comment:
This lemma statement has some redundancy in it. Is there a reason not to perform the substitutions in the statement?
```suggestion
lemma eval_from_split [fintype σ] (x : list α) (s t : σ) :
  fintype.card σ + 1 ≤ x.length →
      ∃ a b c,
        x = a ++ b ++ c ∧
        a.length + b.length ≤ fintype.card σ + 1 ∧
        b ≠ [] ∧
        M.eval_from (M.eval_from s a) b = M.eval_from s a ∧
        M.eval_from (M.eval_from s a) c = M.eval_from s x :=
```
The proof remains basically the same, except a few pieces become unecessary.

If you're looking to avoid repeating `M.eval_from s a`, you can use

```suggestion
lemma eval_from_split [fintype σ] (x : list α) (s t : σ) :
  fintype.card σ + 1 ≤ x.length →
      ∃ a b c,
        x = a ++ b ++ c ∧
        a.length + b.length ≤ fintype.card σ + 1 ∧
        b ≠ [] ∧ let q := M.eval_from s a in
        M.eval_from q b = q ∧
        M.eval_from q c = M.eval_from s x :=
```

Reply:
I feel that this statement hides what the lemma is actually doing, in general, I don't think we really care about particular states of a machine.

Reply:
How about:
```suggestion
lemma eval_from_split [fintype σ] (x : list α) (s : σ) :
  fintype.card σ + 1 ≤ x.length →
    let t = M.eval_from s x in
      ∃ a b c,
        x = a ++ b ++ c ∧
        a.length + b.length ≤ fintype.card σ + 1 ∧
        b ≠ [] ∧ let q = M.eval_from s a in
        M.eval_from q b = q ∧
        M.eval_from q c = t :=
```
It seems odd to say `∃ q, p q ∧ q = x` when you could just write `p x`, which you also do here for `t`.

Let's see what @gebner  thinks

Reply:
The reason I suggested this formulation is because I think of `M.eval_from t x = s` as a ternary relation.  On paper you'd maybe write `t →ˣ s` or `(t,x,s) ∈ δ`.
And the transition relation is really the natural way to look at this lemma: we split `t →ˣ s` into `t →ᵃ q →ᵇ q →ᶜ s`.

I don't think removing one existential quantifier makes this lemma simpler, because there are still existential quantifiers left.  In either case, you're going to use obtain.  And then, if you want fewer equations, you can just write `rfl` in the obtain pattern.

Reply:
Here's yet another suggestion, just making some variables implicit and naming some hypotheses so you can delete `intros` below:
```suggestion
lemma eval_from_split [fintype σ] {x : list α} {s t : σ}
  (hlen : fintype.card σ + 1 ≤ x.length)
  (hx : M.eval_from s x = t) :
  ∃ q a b c,
    x = a ++ b ++ c ∧
    a.length + b.length ≤ fintype.card σ + 1 ∧
    b ≠ [] ∧
    M.eval_from s a = q ∧
    M.eval_from q b = q ∧
    M.eval_from q c = t :=
```